dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -197,32 +198,64 @@ namespace winrt::Microsoft::ProjectReunion::implementation
         check_hresult(UnregisterSystemIdleStatusChangedListener(
             make_self<factory_implementation::PowerManager>()->m_systemIdleStatusHandle));
     }
-   
-    // PowerSchemePersonality Functions
-    EventType& PowerSchemePersonality_Event()
+
+    bool IsEffectivePowerModeV2Supported()
+    {
+        // We check for the version supported by checking if the feature is supported
+        // Using NULL as an indicator for uninitialized cache value
+        if (g_powerModeVersion == NULL)
+        {
+            PVOID handle;
+            auto hr = PowerRegisterForEffectivePowerModeNotifications(
+                EFFECTIVE_POWER_MODE_V2, [](EFFECTIVE_POWER_MODE, PVOID) {}, NULL, &handle);
+            g_powerModeVersion = hr == S_OK ? EFFECTIVE_POWER_MODE_V2 : EFFECTIVE_POWER_MODE_V1;
+            check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(handle));
+        }
+        return g_powerModeVersion == EFFECTIVE_POWER_MODE_V2;
+    }
+
+    EventType& EffectivePowerMode_Event()
     {
-        return make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityChangedEvent;
+        return make_self<factory_implementation::PowerManager>()->m_powerModeChangedEvent;
     }
 
-    void PowerSchemePersonality_Register()
+    void EffectivePowerMode_Register()
     {
-        check_hresult(RegisterPowerSchemePersonalityChangedListener(
-            &PowerManager::PowerSchemePersonalityChanged_Callback,
-            &make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityHandle));
+        ULONG version = IsEffectivePowerModeV2Supported() ? EFFECTIVE_POWER_MODE_V2 : EFFECTIVE_POWER_MODE_V1;
+        check_hresult(PowerRegisterForEffectivePowerModeNotifications(
+            version, [](EFFECTIVE_POWER_MODE mode, PVOID)
+            {
+                PowerManager::EffectivePowerModeChanged_Callback(mode);
+            }, nullptr, &g_powerModeHandle));
     }
 
-    void PowerSchemePersonality_Unregister()
+    void EffectivePowerMode_Unregister()
     {
-        check_hresult(UnregisterPowerSchemePersonalityChangedListener(
-            make_self<factory_implementation::PowerManager>()->m_powerSchemePersonalityHandle));
+        check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(g_powerModeHandle));
+        g_powerModeHandle = NULL;
     }
 
-    void PowerSchemePersonality_Update()
+    void EffectivePowerMode_Update()
     {
-        check_hresult(GetPowerSchemePersonality(
-            &make_self<factory_implementation::PowerManager>()->m_cachedPowerSchemePersonality));
+        // Effectively the Get() for PowerMode
+        // Needs to get a temporary subscription to get most recent value
+        static std::promise<EFFECTIVE_POWER_MODE> promiseObj;
```

Comment:
Forcing a notification roundtrip every time doesn't seem ideal. Assuming that every time somebody gets a PowerModeChanged event, they immediately check what the new mode is, that means we need two notifications.

In theory, can't we say "somebody has a listener registered" plus "that listener has ever fired" equals "m_cachedPowerMode is already up-to-date"? Then we'd only need to do waiting here if a) no listener is registered, or b) somebody registered a listener and then called Get without waiting for that listener to fire.

Reply:
We already have that. A subscription will fire an initial callback with the latest value. And if a get is called then, that latest value is returned.

Reply:
Right, I forgot that this only gets called if there isn't an active subscription.